### PR TITLE
Contract eth_signTypedData

### PR DESCRIPTION
**Pull request by loredanacirstea**
fixes https://github.com/raiden-network/microraiden/issues/139 (EIP712 proposal)
fixes https://github.com/raiden-network/microraiden/issues/140 (gas cost)
fixes https://github.com/raiden-network/microraiden/issues/86 (string concatenations -> ugly code)

- use `eth_signTypedData`-based logic for balance proof messages instead of `personal_sign`
- remove string concatenation code used to recreate balance proof messages inside the contract
- add & fix tests

**Reasons for not supporting both EIP712 & previous logic for verifying the balance proof:**
- we remove code that is harder to follow and that could hide potential vulnerabilities
- we avoid having to introduce another parameter for specifying which `verifyBalanceProof` function to use when closing a channel.

**Other reasons for implementing EIP712:**
- high gas cost for closing a channel; e.g. cooperative close 194k gas

Populus gas cost estimations now, with `eth_signTypedData`-based logic (might be higher in real life):
test_close_by_receiver 60149
test_close_by_sender 69928 (cooperative close)
test_close_by_sender_challenge_settle_by_receiver 108124
test_close_by_sender_challenge_settle_by_sender 103498

**Be aware that:**
- EIP712 is not standardized yet 
  - however, I do not expect the contract implementation to change in a breaking way
- EIP712 is not supported by Parity (MetaMask supports it https://github.com/ethereum/EIPs/pull/712#issuecomment-336346522)
  - we can reproduce EIP712 message hashing -> Parity signer will show human unreadable bytes.
  - some stats: MetaMask extension - 207,351 users ; Parity extension - 14,143 users; Parity app - unknown

### CODE REVIEW

File: contracts/contracts/RaidenMicroTransferChannels.sol [hunk 1]
```diff
@@ -179,21 +150,13 @@ contract RaidenMicroTransferChannels {
...
-        string memory message_length_string = uintToString(message_length);
-
-        // Prefix the message
-        string memory prefixed_message = concat(prefix, message_length_string);
-        prefixed_message = concat(prefixed_message, message);
-
-
-        // Hash the prefixed message string
-        bytes32 prefixed_message_hash = keccak256(prefixed_message);
+        var message_hash = keccak256(
```
Comment (LefterisJP) [inline contracts/contracts/RaidenMicroTransferChannels.sol]: you and me both know that this should be kept in sync with the function, but just add a comment here for future maintainers that these hashed strings should be in sync with this function's parameters

File: contracts/contracts/test/ECVerifyTest.sol [hunk 2]
```diff
@@ -32,4 +29,23 @@ contract ECVerifyTest {
...
+        address _address,
+        uint32 _value,
+        uint192 _value2,
+        bytes _balance_msg_sig)
+        public
+        constant
+        returns (address)
+    {
+        var hash = keccak256(
+          keccak256('address Address', 'uint32 Value', 'uint192 Value2'),
```
Comment (LefterisJP) [inline contracts/contracts/test/ECVerifyTest.sol]: same coment here

File: contracts/utils/sign.py [hunk 3]
```diff
@@ -24,8 +24,30 @@ def eth_message_hex(msg: str) -> bytes:
...
 
-def sign(data: str, private_key_seed_ascii: str):
-    data = eth_message_hex(data)
+def eth_signed_typed_data_message(types, names, data) -> bytes:
+    """
+    types e.g. ('address', 'uint', ('uint', 32))
+    names e.g. ('receiver', 'block_created', 'balance')
+    data e.g. ('0x5601ea8445a5d96eeebf89a67c4199fbb7a43fbb', 3000, 1000)
+    """
+    assert len(types) == len(data) == len(names), 'Check argument length.'
```
Comment (LefterisJP) [inline contracts/utils/sign.py]: I think the assert message should not read like a comment, but like an error message. So something like:
`'Argument tuple length mismatch'`

File: contracts/tests/utils.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+from utils.sign import eth_signed_typed_data_message
+
+def balance_proof_hash(receiver, block, balance):
+    return eth_signed_typed_data_message(('address', ('uint', 32), ('uint', 192)),
```
Comment (LefterisJP) [inline contracts/tests/utils.py]: Style:

```
   return eth_signed_typed_data_message(
       ('address', ('uint', 32), ('uint', 192)),
       ('receiver', 'block_created', 'balance'),
       (receiver, block, balance)
)
```

File: contracts/tests/test_raidenchannels.py [hunk 5]
```diff
@@ -304,10 +304,15 @@ def test_close_by_receiver(web3, chain, contract, channel):
...
-    balance_msg_sig, addr = sign.check(balance_msg, tester.k1)
-    balance_msg_sig_false, addr = sign.check(balance_msg, tester.k3)
+    balance_message_hash = balance_proof_hash(receiver, open_block_number, balance)
+    balance_msg_sig, addr = sign.check(balance_message_hash, tester.k1)
+    balance_msg_sig_false, addr2 = sign.check(balance_message_hash, tester.k3)
+    assert addr == sender
 
+    contract_verified_address = contract.call().verifyBalanceProof(
+                                        receiver, open_block_number,
+                                        balance, balance_msg_sig)
```
Comment (LefterisJP) [inline contracts/tests/test_raidenchannels.py]: style: Close parentheses in its own line. Each arg in its own line

File: contracts/tests/test_ecverify.py [hunk 6]
```diff
@@ -31,52 +36,88 @@ def test_ecrecover_output(web3, ecverify_test_contract):
...
 
     web3.testing.mine(30)
     with pytest.raises(tester.TransactionFailed):
-        ecverify_test_contract.call().verify_ecrecover_output(hash_prefixed_message, bytearray(), s, v)
+        ecverify_test_contract.call().verify_ecrecover_output(balance_message_hash, bytearray(), s, v)
 
     web3.testing.mine(30)
-    verified_address = ecverify_test_contract.call().verify_ecrecover_output(hash_prefixed_message, r, s, v)
+    verified_address = ecverify_test_contract.call().verify_ecrecover_output(balance_message_hash, r, s, v)
     # assert address == verified_address
```
Comment (LefterisJP) [inline contracts/tests/test_ecverify.py]: remove commented line

File: contracts/tests/test_ecverify.py [hunk 7]
```diff
@@ -16,10 +19,12 @@ def ecverify_test_contract(chain, create_contract):
...
-    hash_prefixed_message = sign.eth_message_hex(message)
-    signed_message, addr = sign.check(message, tester.k0)
+    block = 4804175
+    balance = 22000000000000000000
+
+    balance_message_hash = balance_proof_hash(B, block, balance)
+    balance_msg_sig, addr = sign.check(balance_message_hash, tester.k0)
+    assert addr == A
 
     # r = signed_message[:32];
```
Comment (LefterisJP) [inline contracts/tests/test_ecverify.py]: remove commented lines

**Comment by LefterisJP** [comment]
Left some comments. Once addressed this looks good. 

Finally, getting rid of those ugly string concatenations :)

**Comment by loredanacirstea** [comment]
Comments addressed

**Pull request merged**